[NEMO-7] Intra-TaskGroup pipelining

jeongyooneo opened the pull request:
JIRA: [NEMO-7: Refactor TaskGroupExecutor to a batch/stream-unified execution model](https://issues.apache.org/jira/browse/NEMO-7)

**Major changes:**
* Refactored `TaskGroupExecutor` to a pipelined execution model(Thanks to @sanha)
   * Per-`Task` `InputReader`/`OutputWriter` and `TaskState` management are removed.
   * Per-`Task` `Pipe` acts as a local writer for this `Task` to where it emits output, and a local reader for its children `Task`s from where they receives input.
   * Input is deserialized element-wise, trickle down to the end of `TaskGroup` DAG in a recursive manner. 
       * `iteratorIdToDstTasksMap` maintains source input(data from bounded source or other `TaskGroup`s) to the `Task` DAG that processes this input.
       * `pipeToDstTasksMap` maintains intra-TaskGroup input to the `Task` DAG that processes this input.

<p align="center">
<img src="https://user-images.githubusercontent.com/6097406/36969167-11bdbb90-20a8-11e8-89f8-d3d1c459154c.png" width="400">
</p>

<p align="center">
TaskGroup-wise InputReaders and OutputWriters
</p>

<p align="center">
<img src="https://user-images.githubusercontent.com/6097406/36969232-4517e236-20a8-11e8-93e2-f4e8d8ae8f84.png" width="400">
</p>

<p align="center">
Recursive execution of intra-TaskGroup pipelining
</p>


**Minor changes to note:**
* None

**Tests for the changes:**
* The existing tests are sufficient.

**Other comments:**
* None

johnyangk commented on common/src/main/java/edu/snu/nemo/common/ir/Pipe.java:
```diff
@@ -23,7 +23,7 @@
  * @param <O> output type.
  */
-public interface OutputCollector<O> extends Serializable {
+public interface Pipe<O> extends Serializable {
```
IMHO, `OutputCollector` is a better name.
- `OutputCollector` explicitly tells you that it is used to collect outputs
- People may confuse this `Pipe` class with the `Pipe` execution property.

Storm/Flink also use `OutputCollector` as the name of the class for emitting outputs.
- Storm: https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/task/OutputCollector.java
- Flink: https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputCollector.java

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+        });
       }
+      setTaskPutOnHold((MetricCollectionBarrierTask) task);
     }
```
add else{}, and throw an exception in it.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+        Iterator iterator = idToSrcIteratorMap.get(srcIteratorId);
+        iterator.forEachRemaining(element -> {
+          for (final Task task : tasks) {
+            List data = Collections.singletonList(element);
```
I'd avoid creating an extra object per data element.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+
+    // Process element-wise depending on the Task type
+    if (task instanceof BoundedSourceTask) {
+      LocalPipe pipe = taskToOutputPipeMap.get(task);
```
I'd avoid computing a hash value per data element.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+  }
+
+  private void prepareTransform(final Transform transform, final Task task) {
+    if (!preparedTransforms.contains(transform)) {
```
Ditto on the extra per-element operation.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
-      } // If else, this is a sink task.
+      final Transform.Context transformContext = new ContextImpl(sideInputMap);
+      final LocalPipe outputPipe = taskToOutputPipeMap.get(task);
+      transform.prepare(transformContext, outputPipe);
```
I'd prepare all transforms of a task group, before processing any data.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+   * Executes the task group.
+   */
+  public void execute() {
+    final Map<String, Object> metric = new HashMap<>();
```
Please refactor metric collection code into a separate class, so that its LOC is minimal in `TaskGroupExecutor`.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final Set<Transform> preparedTransforms;
+  private final Set<String> finishedTaskIds;
+  private final AtomicInteger completedFutures;
+  private int numBoundedSources;
```
Can this become a local variable?

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final Set<String> finishedTaskIds;
+  private final AtomicInteger completedFutures;
+  private int numBoundedSources;
+  private int numIterators;
```
Can this become a local variable?

jeongyooneo replied:
Not really, since TaskGroupExecutor should guarantee that it consumed all partitions it received. I'll rename numIterators to numPartitions for clarity.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  // Map of task ID to its intra-TaskGroup data pipe.
+  private final Map<Task, List<LocalPipe>> taskToInputPipesMap;
+  private final Map<Task, LocalPipe> taskToOutputPipeMap;  // one and only one Pipe per task
+  // Readers/writers that deals with inter-TaskGroup data.
```
I'm not sure about the newly added data structures below for the following reasons
- Some of them cost an extra per-element operation and memory overhead
- Maybe we can do without some of them

Alternatively, I'd think about the following options
- Compose iterators like Spark (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/MapPartitionsRDD.scala#L38)
- Reuse `taskGroupDag` to obtain dependency information

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private volatile Map<String, List<Task>> pipeIdToDstTasksMap;
+  private final Set<Transform> preparedTransforms;
+  private final Set<String> finishedTaskIds;
+  private final AtomicInteger completedFutures;
```
Remove? It does `getAndIncrement()`, but the returned value is not used.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> iteratorQueue;
+  private volatile Map<String, List<Task>> pipeIdToDstTasksMap;
+  private final Set<Transform> preparedTransforms;
+  private final Set<String> finishedTaskIds;
```
Can you do without this?
(e.g., We assume a task is finished, when it consumes all of its input iterators)

jeongyooneo replied:
I don't think so. From the point of view of Task, it is not know when it consumed all the iterators. Even at the point when all iterators are received and consumed in `execute()`, there can exist some Tasks at downstream that haven't received it.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final Map<String, List<Task>> iteratorIdToTasksMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> iteratorQueue;
+  private volatile Map<String, List<Task>> pipeIdToDstTasksMap;
+  private final Set<Transform> preparedTransforms;
```
Remove, and instead prepare transforms before consuming any data?

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final Map<String, Iterator> idToSrcIteratorMap;
+  private final Map<String, List<Task>> srcIteratorIdToTasksMap;
+  private final Map<String, List<Task>> iteratorIdToTasksMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> iteratorQueue;
```
Please use a more descriptive name.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private final Map<String, List<Task>> srcIteratorIdToTasksMap;
+  private final Map<String, List<Task>> iteratorIdToTasksMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> iteratorQueue;
+  private volatile Map<String, List<Task>> pipeIdToDstTasksMap;
```
Can you explain in the comment why this is marked as `volatile`? My understanding is that only a single thread instantiates and executes a `TaskGroupExecutor`. Let me know if I'm missing something.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private int numBoundedSources;
+  private int numIterators;
+
+  // For metrics
```
Can you refactor these variables into a separate class?

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private volatile boolean isExecutionRequested;
+  private volatile String logicalTaskIdPutOnHold;
+
+  private static final String ITERATORID_PREFIX = "ITERATOR_";
```
Can you do without iterator ids?
(e.g., composing iterators, using pointers in a DAG)

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  // For metrics
+  private long serBlockSize;
+  private long encodedBlockSize;
+  private long accumulatedBlockedReadTime;
```
Remove unused variable.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+            getPhysicalTaskId(parent.getId()), taskGroupId, physicalTaskId);
+      });
+      taskToInputPipesMap.put(task, inputPipes);
+    }
```
add else {} and add a comment in it to explain why we're not handling that case..

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
+    final List<RuntimeEdge<Task>> outEdges = taskGroupDag.getOutgoingEdgesOf(task);
+
+    outEdges.forEach(outEdge -> {
+      if (outEdge.isSideInput()) {
```
Ditto on else{}.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private long encodedBlockSize;
+  private long accumulatedBlockedReadTime;
+
+  private volatile boolean isExecutionRequested;
```
Ditto on volatile.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,36 @@
+  private long accumulatedBlockedReadTime;
+
+  private volatile boolean isExecutionRequested;
+  private volatile String logicalTaskIdPutOnHold;
```
Ditto on volatile.

johnyangk commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
-      final Optional<Long> writtenBytes = outputWriter.getWrittenBytes();
-      writtenBytes.ifPresent(writtenBytesList::add);
+    // For the produced output
+    LocalPipe pipe = taskToOutputPipeMap.get(task);
```
Ditto on per-element operation.

johnyangk commented:
@jeongyooneo In general, it'd be great if an element directly travels Transform -> List<Object>(i.e., OutputCollector) -> Transform -> ..., without computing a hash or creating a new object in the process.

seojangho commented on common/src/main/java/edu/snu/nemo/common/ir/vertex/transform/Transform.java:
```diff
@@ -32,16 +30,15 @@
+   * @param element data received.
    */
-  void onData(Iterator<I> elements, String srcVertexId);
+  void onData(Object element);
```
We declared input type as `<I>`.
```java
void onData(I element);
```

seojangho commented on common/src/main/java/edu/snu/nemo/common/test/ArgBuilder.java:
```diff
@@ -20,11 +20,14 @@
  * Argument builder.
  */
 public final class ArgBuilder {
+  private static final List<List<String>> DEFAULT_ARGS = Arrays.asList(Arrays.asList("-executor_json",
```
These changes may conflict with `master`. Please follow `master` to resolve this.

seojangho commented on compiler/frontend/beam/src/main/java/edu/snu/nemo/compiler/frontend/beam/transform/FlattenTransform.java:
```diff
@@ -37,18 +36,18 @@ public FlattenTransform() {
   @Override
-  public void onData(final Iterator<T> elements, final String srcVertexId) {
-    elements.forEachRemaining(collectedElements::add);
+  public void onData(final Object element) {
```
Same here. `public void onData(final T element)`?

seojangho commented on compiler/frontend/spark/src/main/java/edu/snu/nemo/compiler/frontend/spark/transform/GroupByKeyTransform.java:
```diff
@@ -38,22 +38,23 @@ public GroupByKeyTransform() {
-      keyToValues.get(element._1).add(element._2);
-    });
+  public void onData(final Object element) {
+    K key = ((Tuple2<K, V>) element)._1;
```
Add `final` modifier?

seojangho commented on compiler/frontend/spark/src/main/java/edu/snu/nemo/compiler/frontend/spark/transform/ReduceByKeyTransform.java:
```diff
@@ -42,24 +46,29 @@ public ReduceByKeyTransform(final Function2<V, V, V> func) {
-      keyToValues.get(element._1).add(element._2);
-    });
+  public void onData(final Object element) {
+    K key = ((Tuple2<K, V>) element)._1;
```
Add `final` modifier?

seojangho commented on compiler/frontend/spark/src/main/java/edu/snu/nemo/compiler/frontend/spark/transform/ReduceTransform.java:
```diff
@@ -15,41 +15,62 @@
-  public void onData(final Iterator<T> elements, final String srcVertexId) {
-    final T res = reduceIterator(elements, func);
-    if (res == null) { // nothing to be done.
+  public void onData(final Object element) {
```
I suggest `public void onData(final T element)`

seojangho commented on conf/src/main/java/edu/snu/nemo/conf/JobConf.java:
```diff
@@ -103,7 +103,7 @@
    */
   @NamedParameter(doc = "Path to the JSON file that specifies resources for executors", short_name = "executor_json",
-  default_value = "examples/resources/sample_executor_resources.json")
+  default_value = "../resources/sample_executor_resources.json")
```
This change may conflict with `master`. Please follow `master`.

sanha commented:
For the @johnyangk's comment (avoiding hashing and creating objects), I'd suggest the next model.
- Build a DAG of `TaskWrapper` (or something like that) from `taskGroupDag` of `ScheduledTaskGroup` when a `TaskGroup` is scheduled.
  - This DAG should manage the connection among vertices as pointer (reference) rather than `Map` and `List`, unlike our current `DAG` implementation.
  - The `TaskWrapper` should have `Callable`, which consumes input element and produce output.
    - This `Callable` can be built from the `Transform` of `Task` that the wrapper wraps.
  - The `TaskWrapper` can have any other stuffs which are stored in `TaskDataHandler` now.
- After this, each data in input data `Iterable` can be processed through this DAG of `TaskWrapper` without calculating any hash or creating any extra object.

sanha commented:
Unfortunately, I've noticed that there is a critical bug in this change.
When I run our `MapReduce` example for 512 GB Wikimedia data in Amazon EC2, all executors hanged at the first `TaskGroup` of reduce stage. (It seems that this hanging does not occur in our ITCase size.)
We must resolve this problem before we merge this PR. Let's investigate this issue thoroughly.

sanha commented:
I strongly suspect that the sink vertex creates temporary file per every element.
After I killed the hanged job, HDFS failed to access the output directory.

bgchun commented:
@jeongyooneo How do you test your changes in this PR?

sanha commented:
Here's a good news! 
I confirmed that the bug I've reported is fixed after the last commit.
The experiment works well. Thanks for the hard work!

jeongyooneo commented:
@bgchun I’ve modified the ‘TaskGroupExecutor’ to element-wise version and checked that all ITCases worked correctly with this PR. 
If you have any additional suggestion, please let me know, and I’ll try it right away!

jeongyooneo commented:
Thanks for the review! 
I've addressed most of them, except the following two:
1) For refactoring metric collection code into a separate class requested by @johnyangk, I think we'd better do that in another issue. 
Since there are more types of metric to be added, it'd be great if we clean them up after the types of metric we need and their usages become concrete.
2) For checking TaskGroup complete event by collecting finished Task ids, there are subtle task completion timing issues that made me keep this as is.
- TaskGroup is complete when all of its Tasks are complete, i.e. each Task consumed all of its input.
- As input is processed element-wise, speculating the exact point when _final_ task is finished is hard. There can be multiple _final_ tasks that process input independently, whose finishing point are known only to each task. Thus, we still need to collect those _final_ task ids to check TaskGroup completion.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,13 +55,27 @@
-   */
-  private final Map<String, List<InputReader>> physicalTaskIdToInputReaderMap;
-  private final Map<String, List<OutputWriter>> physicalTaskIdToOutputWriterMap;
+  private final List<InputReader> inputReaders;
```
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,14 +52,36 @@
+  private final Map<String, List<TaskDataHandler>> srcIteratorIdToDataHandlersMap;
+  private final Map<String, List<TaskDataHandler>> iteratorIdToDataHandlersMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> partitionQueue;
+  private List<TaskDataHandler> taskDataHandlers;
```
Do we have to maintain this list and `srcIteratorIdToDataHandlersMap`, `srcIteratorIdToDataHandlersMap` at the same time?

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -60,14 +52,36 @@
+  private final Map<String, List<TaskDataHandler>> iteratorIdToDataHandlersMap;
+  private final LinkedBlockingQueue<Pair<String, DataUtil.IteratorWithNumBytes>> partitionQueue;
+  private List<TaskDataHandler> taskDataHandlers;
+  private Map<OutputCollectorImpl, List<TaskDataHandler>> outputToChildrenDataHandlersMap;
```
Can't we see this information in `TaskDataHandler`?

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
-    final String physicalTaskId = getPhysicalTaskId(boundedSourceTask.getId());
-    final Map<String, Object> metric = new HashMap<>();
-    metricCollector.beginMeasurement(physicalTaskId, metric);
+  private boolean hasOutputWriter(final Task task) {
```
Please add comments to all these private methods.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+        } catch (final BlockFetchException ex) {
+          taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.FAILED_RECOVERABLE,
+              Optional.empty(), Optional.of(TaskGroupState.RecoverableFailureCause.INPUT_READ_FAILURE));
+          LOG.info("{} Execution Failed (Recoverable: input read failure)! Exception: {}",
```
`LOG.error` ?

sanha replied:
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+        } catch (final Exception e) {
+          taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.FAILED_UNRECOVERABLE,
+              Optional.empty(), Optional.empty());
+          LOG.info("{} Execution Failed! Exception: {}", taskGroupId, e.toString());
```
ditto

sanha replied:
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+          throw new RuntimeException(e);
+        }
+      }
+      // TODO #XXX: Support other types of source tasks, i. e. InitializedSourceTask
```
Check this TODO. If it is needed, please create the issue and specify the issue number.

sanha replied:
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+          serBlockSize += iterator.getNumSerializedBytes();
+        } catch (final DataUtil.IteratorWithNumBytes.NumBytesNotSupportedException e) {
+          serBlockSize = -1;
+        } catch (final IllegalStateException e) {
```
When it throws `IllegalStateException`?

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+        } catch (final DataUtil.IteratorWithNumBytes.NumBytesNotSupportedException e) {
+          encodedBlockSize = -1;
+        } catch (final IllegalStateException e) {
+          LOG.error("IllegalState ", e);
```
ditto

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+            final Object element = outputCollector.remove();
+
+            // Pass outputCollectorOwnerTask's output to its children tasks recursively.
+            if (!childrenDataHandlers.isEmpty()) {
```
do we have to check this?

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+    } catch (final BlockWriteException ex2) {
+      taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.FAILED_RECOVERABLE,
+          Optional.empty(), Optional.of(TaskGroupState.RecoverableFailureCause.OUTPUT_WRITE_FAILURE));
+      LOG.info("{} Execution Failed (Recoverable: output write failure)! Exception: {}",
```
`LOG.error`?

sanha replied:
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
+    } catch (final Exception e) {
+      taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.FAILED_UNRECOVERABLE,
+          Optional.empty(), Optional.empty());
+      LOG.info("{} Execution Failed! Exception: {}",
```
ditto

sanha replied:
Please check this comment again.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java:
```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
-      } catch (final InterruptedException e) {
-        throw new BlockFetchException(e);
+    // Process element-wise depending on the Task type
+    if (task instanceof BoundedSourceTask) {
```
If we pre-build these stuffs as `Callable` in `TaskDataHandler`, we can avoid to doing `instanceof`s per every elements.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/OutputCollectorImpl.java:
```diff
@@ -46,11 +61,39 @@ public void emit(final String dstVertexId, final Object output) {
    */
-  public List<O> collectOutputList() {
-    return outputList.getAndSet(new ArrayList<>());
+  public O remove() {
```
Please check this comment again

johnyangk commented:
Thanks @sanha. I'll make another pass once @jeongyooneo addresses the comments here.

jeongyooneo commented:
@sanha I've addressed most of the comments. The following two I haven't resolved will be addressed via another issue of refactoring `TaskDataHandler`-based data transfer:
- Removing local variables dealing with iterators in `TaskGroupExecutor`
- Pre-build `runTask` per-task type operation as Callable in TaskDataHandler

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/OutputCollectorImpl.java:
```diff
@@ -85,15 +83,16 @@ public RuntimeEdge getSideInputRuntimeEdge() {
     sideInputReceivers.add(physicalTaskId);
   }
 
   public boolean hasSideInputFor(final String physicalTaskId) {
```
Comment please :)

sanha commented on compiler/frontend/spark/src/main/java/edu/snu/nemo/compiler/frontend/spark/transform/CollectTransform.java:
```diff
@@ -22,44 +22,49 @@
  */
 public final class CollectTransform<T> implements Transform<T, T> {
   private String filename;
+  private FileOutputStream fos;
```
Why do we need to have these `fos` and `oos` as variables? It can be created only in `close()`.

sanha commented on compiler/frontend/spark/src/main/java/edu/snu/nemo/compiler/frontend/spark/transform/CollectTransform.java:
```diff
@@ -22,44 +22,49 @@
 
   @Override
   public void close() {
+    try {
```
Please use try-with-resource clause for these `Stream`s.

sanha commented on compiler/optimizer/src/main/java/edu/snu/nemo/compiler/optimizer/examples/EmptyComponents.java:
```diff
@@ -60,7 +59,7 @@ public void prepare(final Context context, final OutputCollector<O> outputCollec
 
     @Override
-    public void onData(final Iterator<I> elements, final String srcVertexId) {
+    public void onData(final Object element) {
```
`I element`?

sanha commented on examples/beam/src/test/java/edu/snu/nemo/examples/beam/policy/DataSkewPolicyParallelismFive.java:
```diff
@@ -25,10 +25,10 @@
  * A data-skew policy with fixed parallelism 5 for tests.
  */
-public final class DataSkewPolicyParallelsimFive implements Policy {
+public final class DataSkewPolicyParallelismFive implements Policy{
```
Please revert the space between `Policy` and bracket.

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/OutputCollectorImpl.java:
```diff
@@ -46,11 +52,47 @@ public void emit(final String dstVertexId, final Object output) {
+    sideInputRuntimeEdge = edge;
+  }
+
+  public RuntimeEdge getSideInputRuntimeEdge() {
```
Please add a comment :)

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/OutputCollectorImpl.java:
```diff
@@ -46,11 +52,47 @@ public void emit(final String dstVertexId, final Object output) {
+  public void setAsSideInputFor(final String physicalTaskId) {
+    sideInputReceivers.add(physicalTaskId);
+  }
+
```
Please add a comment :)

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/OutputWriter.java:
```diff
@@ -69,11 +69,19 @@ public OutputWriter(final int hashRangeMultiplier,
+   * Collects output element-wise in memory.
    *
-   * @param dataToWrite An iterable for the elements to be written.
+   * * @param element
```
Please remove the duplicated *

sanha commented on runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/datatransfer/TaskDataHandler.java:
```diff
@@ -0,0 +1,94 @@
+public final class TaskDataHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TaskDataHandler.class.getName());
+
+  public TaskDataHandler(final Task task) {
```
Please add comments for the constructor and getters.

jeongyooneo commented:
@sanha Just finished addressing comments!
